pkg/cmd/server/apis/config/v1/types.go

```diff
@@ -770,6 +774,20 @@ type OAuthTemplates struct {
 	Error string `json:"error"`
 }
 
+// ExternalOAuthConfig allos to configure external OAuth server discovery
+type ExternalOAuthConfig struct {
+	// MetadataFile is a path to a file containing the discovery endpoint for OAuth 2.0 Authorization Server Metadata
+	// for an External OAuth server.
+	// See IETF Draft: // https://tools.ietf.org/html/draft-ietf-oauth-discovery-04#section-2
+	MetadataFile string `json:"metadataFile"`
+
+	// MasterPublicURL is used for building valid client redirect URLs for internal and external access
+	MasterPublicURL string `json:"masterPublicURL"`
```

Comment:
This is used to...
 1. set the authorization URL for openapi. This usage is incorrect, it should be derivable from the metadata file above
 2. the metadata file we "normally" serve.  You should update our code to always serve from a metadata file.  You have the information to build it.
 3. an https check you don't need.  It's always https
 4. a non-standard display-token endpoint that is only correct when using the embedded oauth server
 5. a /token/implicit endpoint, which I don't recall.  Might have used it to terminate the CLI flow?  An external provider should set up the valid URL themselves.
 6. an authorize endpoint that actually points to the oauthserver.  discoverable from metadatafile.
 7. a non-standard token request endpoint only valid for the embedded oauth server
 8. the non-standard display-token endpoint itself, which is part of the embedded oauth server

Can you help me find the use-case that requires this field?

Reply:
1.  I can change that to use GetMasterURL, but the whoe openapi thing is generated only when we enable the OAuth Server. should we generate the openapi stuff entirely from the metdata file instead ? Who/What uses this openapi stuff anyway ?

Reply:
2. sounds like chicken and egg here: the metadata we generate uses that OAuthConfig.MasterPublicURL for generation, so where do we get the value from initially ?

Reply:
3. We do not force OAuthConfig.MAsterPublicURL to be https

Reply:
4. This is all dependent on OAuthConfig and using that configuration, nothing to do with ExternlOAuthConfig

Reply:
5. I guess I can unravel this my using the Issuer from the metadata

Reply:
On 4 again, I see what you mean, the only way I can deal with this is to use Issuer there too ... even though it may be a dead endopint when an External OAuth Server is used

Reply:
After some discussion with @enj I think I can avoid bootstrapping the oauthclients for external oauth servers. And in the worst case we can use an ansible playbook to populate oauthclients for an external oauth server that needs/wants them in openshift. So I am going to backtrack some of this and ping you later